packages/vm/test/api/EIPs/eip-3855.spec.ts

```diff
@@ -19,14 +19,11 @@ describe('EIP 3541 tests', () => {
     const vm = await VM.create({ common })
     let stack: bigint[]
     vm.evm.events!.on('step', (e: InterpreterStep) => {
-      if (typeof stack !== 'undefined') {
-        assert.fail('should only do PUSH0 once')
-      }
       stack = e.stack
     })
 
     const result = await vm.evm.runCode!({
-      code: hexToBytes('0x5F'),
+      code: hexToBytes('0x5F00'),
```

Comment:
Note: this test had to change because before this PR, the `stack` item in this test is a pointer to the actual internal stack of the EVM. Once `step` is called, the stack is empty. Then, PUSH0 is used and thus 0 is pushed on stack. Thus the pointer to stack is now `[0]`. However, with the new change, we make a copy of the stack to represent the actual state of the stack. Thus, at the step and also after the step, it will not change and stay empty.

If we instead use `STOP` (opcode `0x00`) then the stack itself will thus change to the stack *before* using STOP (thus `[0]`), which fixes this test.

Note: this PR will thus make the `step` event **much** slower since we now copy the stack instead of reporting a pointer to it.

Reply:
Actually that was/would be a bit my suggestion, that we fall back to the old behavior once someone is listening to the step event and then one time correct the stack to the correct height and then do the respective pushes and pops.

A bit more code to handle. Would this make sense? 🤔

Reply:
Yes, this makes sense. So I will introduce `PerformanceStack` and `OldStack` (will think of better names) and will use `OldStack` if we have an event listener.

There is one edge case, what if we are already running EVM and then we start listening to the step event?

Reply:
I don't think this will work too separate this into two stacks. This is not an edge case you mentioned, listeners can come and go all the time during runtime.

But I think it should be possible to do in one integrated stack implementation?

Reply:
Ok, will nevertheless merge here for now.